libs/header/p2p/session.go

```diff
@@ -139,53 +147,85 @@ func (s *session[H]) doRequest(
 
 	r, size, duration, err := sendMessage(ctx, s.host, stat.peerID, s.protocolID, req)
 	if err != nil {
-		log.Errorw("requesting headers from peer failed", "failed peer", stat.peerID, "err", err)
-		select {
-		case <-s.ctx.Done():
-		case s.reqCh <- req:
-			stat.decreaseScore()
-			log.Debug("retrying the request from different peer")
-		}
-		return
+		// we should not punish peer at this point and should try to parse responses, despite that error was received.
+		log.Debugw("requesting headers from peer failed", "peer", stat.peerID, "err", err)
 	}
 
 	h, err := s.processResponse(r)
 	if err != nil {
 		switch err {
-		case header.ErrNotFound:
+		case header.ErrNotFound, errEmptyResponse:
+			stat.decreaseScore()
 		default:
 			s.peerTracker.blockPeer(stat.peerID, err)
 		}
+
+		// exclude header.ErrNotFound from being logged as it is a `valid` error
```

Comment:
a lot of this could be avoided if we eagerly exchanged HELLO messages with peers as we connect to them that contain their Head :)

Reply:
This doesn't solve potential issues when we enable pruning